### PR TITLE
Remove a bunch of allocations

### PR DESCRIPTION
**Pull request by Maxwe11**

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/UsingDirectiveSyntaxHelpers.cs [hunk 1]
```diff
@@ -15,7 +15,13 @@ internal static class UsingDirectiveSyntaxHelpers
 
         private static bool ExcludeGlobalKeyword(IdentifierNameSyntax token) => !token.Identifier.IsKind(SyntaxKind.GlobalKeyword);
 
-        private static SyntaxToken? GetFirstIdentifierInUsingDirective(UsingDirectiveSyntax usingDirective) => usingDirective.DescendantNodes().OfType<IdentifierNameSyntax>().FirstOrDefault(ExcludeGlobalKeyword)?.Identifier;
+        private static SyntaxToken? GetFirstIdentifierInUsingDirective(UsingDirectiveSyntax usingDirective)
+        {
+            return usingDirective.DescendantNodes()
+                .OfType<IdentifierNameSyntax>()
+                .FirstOrDefault(x => ExcludeGlobalKeyword(x))?
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/UsingDirectiveSyntaxHelpers.cs]: :exclamation: Don't split the `?` and `:`.

Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/UsingDirectiveSyntaxHelpers.cs]: fixed

**Comment by sharwell** [comment]
I like many of the changes you made because they make the code cleaner. For items that don't make the code cleaner, I'd rather not make the changes unless they are showing up on profiling reports. I'll add comments only on the ones I'm hesitant about.

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/UsingDirectiveSyntaxHelpers.cs [hunk 2]
```diff
@@ -15,7 +15,12 @@ internal static class UsingDirectiveSyntaxHelpers
 
         private static bool ExcludeGlobalKeyword(IdentifierNameSyntax token) => !token.Identifier.IsKind(SyntaxKind.GlobalKeyword);
 
-        private static SyntaxToken? GetFirstIdentifierInUsingDirective(UsingDirectiveSyntax usingDirective) => usingDirective.DescendantNodes().OfType<IdentifierNameSyntax>().FirstOrDefault(ExcludeGlobalKeyword)?.Identifier;
+        private static SyntaxToken? GetFirstIdentifierInUsingDirective(UsingDirectiveSyntax usingDirective)
+        {
+            return usingDirective.DescendantNodes()
+                .OfType<IdentifierNameSyntax>()
+                .FirstOrDefault(x => ExcludeGlobalKeyword(x))?.Identifier;
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/UsingDirectiveSyntaxHelpers.cs]: :bulb: Move `?.Identifier` to the next line.

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/XmlCommentHelper.cs [hunk 3]
```diff
@@ -171,7 +171,7 @@ internal static string GetText(XmlTextSyntax textElement, bool normalizeWhitespa
 
             foreach (var item in textElement.TextTokens)
             {
-                stringBuilder.Append(item);
+                stringBuilder.Append(item.ToString());
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/XmlCommentHelper.cs]: :grey_exclamation: This doesn't make the code cleaner.

File: StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1304SA1311CodeFixProvider.cs [hunk 4]
```diff
@@ -56,7 +56,7 @@ public override FixAllProvider GetFixAllProvider()
                     continue;
                 }
 
-                var newName = char.ToUpper(token.ValueText[0]) + token.ValueText.Substring(1);
+                var newName = char.ToUpper(token.ValueText[0]).ToString() + token.ValueText.Substring(1);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1304SA1311CodeFixProvider.cs]: :grey_exclamation: This doesn't make the code cleaner.

File: StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1306CodeFixProvider.cs [hunk 5]
```diff
@@ -52,7 +52,7 @@ public override FixAllProvider GetFixAllProvider()
 
                 if (!string.IsNullOrEmpty(token.ValueText))
                 {
-                    var newName = char.ToLower(token.ValueText[0]) + token.ValueText.Substring(1);
+                    var newName = char.ToLower(token.ValueText[0]).ToString() + token.ValueText.Substring(1);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1306CodeFixProvider.cs]: :grey_exclamation: This doesn't make the code cleaner.

File: StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1307CodeFixProvider.cs [hunk 6]
```diff
@@ -51,7 +51,7 @@ public override FixAllProvider GetFixAllProvider()
 
                 if (!string.IsNullOrEmpty(token.ValueText))
                 {
-                    var newName = char.ToUpper(token.ValueText[0]) + token.ValueText.Substring(1);
+                    var newName = char.ToUpper(token.ValueText[0]).ToString() + token.ValueText.Substring(1);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1307CodeFixProvider.cs]: :grey_exclamation: This doesn't make the code cleaner.

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1120CommentsMustContainText.cs [hunk 7]
```diff
@@ -66,12 +66,10 @@ private static void HandleSyntaxTree(SyntaxTreeAnalysisContext context)
...
         private static void HandleMultiLineComment(SyntaxTreeAnalysisContext context, SyntaxTrivia multiLineComment)
         {
-            var nodeText = multiLineComment.ToString();
+            var commentText = multiLineComment.ToString();
 
-            // We remove the /* and the */ and determine if the comment has any content.
-            var commentText = nodeText.Substring(2, nodeText.Length - 4);
-
-            if (string.IsNullOrWhiteSpace(commentText))
+            // We and determine if the comment has any content between /* and */
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1120CommentsMustContainText.cs]: :exclamation: The new wording doesn't make sense. Perhaps you meant to leave out the first 'and'?

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1120CommentsMustContainText.cs [hunk 8]
```diff
@@ -99,21 +97,22 @@ private static void HandleSingleLineComment(SyntaxTreeAnalysisContext context, S
...
-        private static bool IsNullOrWhiteSpace(string value, int startIndex)
+        private static bool IsNullOrWhiteSpace(string value, int startIndex, int count)
         {
             if (value == null)
             {
                 return true;
             }
 
-            for (int i = startIndex; i < value.Length; i++)
+            for (int i = startIndex, length = startIndex + count; i < length; i++)
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1120CommentsMustContainText.cs]: :bulb: Consider moving the initialization of `length` to the previous line.

**Comment by pdelvo** [comment]
This PR is not mergeable anymore. @Maxwe11 Do you want to update it or should we close it?

**Pull request closed without merging**